### ISSUE TITLE
angular 7: Almost empty AppModule is producing huge main.js (= 367kb)

### ISSUE THREAD

**Issue by goodmite** [angular/angular#32808]
# 🐞 bug report
I have an app with almost empty `AppModule`. After `ng build --prod` I am getting `main.js` size  = 367kb. Expected is around 100kb.

### Is this a regression?
No

### Description
I have an Angular 7 application consisting of 150+ components. I am trying to reduce first load time of this app by reducing the size of` main.js`. I removed all declarations and imports from `AppModule `except `AppComponent `and served the app. This gave me a main.js of size 367kb.
I tried using `webpack-bundle-analyser` and here is the results for `main.js`

<img width="960" alt="github" src="https://user-images.githubusercontent.com/16303528/65390882-2e121480-dd80-11e9-8327-9a7d97d49586.png">

I have been struggling with this issue with the last 10 months. I have wasted countless hours and tried given up many times. I have posted on Reddit about it but no help. So I am posting it here.

## 🔬 Minimal Reproduction
Git hub repo: https://github.com/goodmite/angular-hello-world-size-too-big
To serve: `ng s`
visualize using `webpack-bundle-analyser`:  `npm run bundle-report`

## 🌍  Your Environment

**Angular Version:**
<pre><code>
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 7.3.9
Node: 10.16.3
OS: win32 x64
Angular: 7.2.15
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, platform-server, router
... service-worker

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.10.2
@angular-devkit/build-angular     0.10.2
@angular-devkit/build-optimizer   0.10.2
@angular-devkit/build-webpack     0.10.2
@angular-devkit/core              0.7.5
@angular-devkit/schematics        0.7.5
@angular/cdk                      7.3.7
@angular/cli                      7.3.9
@angular/material                 7.3.7
@angular/pwa                      0.7.5
@ngtools/webpack                  7.0.2
@schematics/angular               0.7.5
@schematics/update                0.13.9
rxjs                              6.5.2
typescript                        3.1.6
webpack                           4.19.1

</code></pre>

**Comment by Toxicable** [comment]
on a new prject im seeing a main size of about 245kb
`yarn init . && yarn install @angular/cli && ng new size && cd size && ng build --prod --stats-json && npx webpack-bundle-analyzer dist/size/stats-es2015.json`

**Comment by trotyl** [comment]
[Result in CI](https://circleci.com/gh/trotyl/ng-size-tracking/19):

```
-rw-r--r-- 1 circleci  15536 Sep 23 04:29 3rdpartylicenses.txt
-rw-r--r-- 1 circleci   4652 Sep 23 04:29 3rdpartylicenses.txt.gz
-rw-r--r-- 1 circleci    948 Sep 23 04:29 favicon.ico
-rw-r--r-- 1 circleci    983 Sep 23 04:29 favicon.ico.gz
-rw-r--r-- 1 circleci    810 Sep 23 04:30 index.html
-rw-r--r-- 1 circleci    405 Sep 23 04:30 index.html.gz
-rw-r--r-- 1 circleci 173212 Sep 23 04:30 main-es2015.63b3a7731cf54ad62923.js
-rw-r--r-- 1 circleci  51214 Sep 23 04:30 main-es2015.63b3a7731cf54ad62923.js.gz
-rw-r--r-- 1 circleci 187901 Sep 23 04:30 main-es5.63b3a7731cf54ad62923.js
-rw-r--r-- 1 circleci  53430 Sep 23 04:30 main-es5.63b3a7731cf54ad62923.js.gz
-rw-r--r-- 1 circleci  37277 Sep 23 04:29 polyfills-es2015.0fe6949bc5ff4b784062.js
-rw-r--r-- 1 circleci  12540 Sep 23 04:29 polyfills-es2015.0fe6949bc5ff4b784062.js.gz
-rw-r--r-- 1 circleci 124853 Sep 23 04:30 polyfills-es5.a83ac866abc867bfd530.js
-rw-r--r-- 1 circleci  42593 Sep 23 04:30 polyfills-es5.a83ac866abc867bfd530.js.gz
-rw-r--r-- 1 circleci   1485 Sep 23 04:29 runtime-es2015.85f895af57b038f1e5b4.js
-rw-r--r-- 1 circleci    758 Sep 23 04:29 runtime-es2015.85f895af57b038f1e5b4.js.gz
-rw-r--r-- 1 circleci   1485 Sep 23 04:29 runtime-es5.85f895af57b038f1e5b4.js
-rw-r--r-- 1 circleci    755 Sep 23 04:29 runtime-es5.85f895af57b038f1e5b4.js.gz
-rw-r--r-- 1 circleci      0 Sep 23 04:29 styles.3ff695c00d717f2d2a11.css
-rw-r--r-- 1 circleci     52 Sep 23 04:29 styles.3ff695c00d717f2d2a11.css.gz
```

`main.js` is only 169.2 KB (50.0KB Gzipped).

Unless you can provide a fresh reproduction (has nothing to do with your existing project), it is a support question and should not be opened here.

**Comment by goodmite** [comment]
I updated the [project on github](https://github.com/goodmite/angular-hello-world-size-too-big) to contain only three modules and removed everything else.
Following is the app Architecture:
<img width="487" alt="gh" src="https://user-images.githubusercontent.com/16303528/65441841-b825b000-de48-11e9-9f6d-c42fc70dd734.png">

As you can see here in [`TestModule`](https://github.com/goodmite/angular-hello-world-size-too-big/blob/master/src/app/core/test.module.ts), I have imported 4 modules for demonstation purposes (`RouterModule`, `HttpClientModule`, `FormsModule`, `ReactiveFormsModule`)

If you comment these 4 modules, the main.js size is 182kb (= 51kb gzip)
If you uncomment these 4 modules, the main.js size is 195kb (= 54.6kb gzip)

**Problem:** 
Since `TestModule `is not imported in `AppModule `in any way, adding or removing modules
     in `TestModule` should have no impact on main.js file size

**Comment by trotyl** [comment]
> I updated the project on github to contain only three modules and removed everything else.

@goodmite Instead of patching your project, you should create a newly generated Angular CLI project to reproduce the issue, no one is likely to check your workspace to get out every nuance.

**Comment by goodmite** [comment]
@trotyl Its a minimal project to reproduce the problem. All 3 modules are almost empty.

But anyway, as per your advice, I have created a fresh angular 8 project and this issue is reproducible too.

**Comment by trotyl** [comment]
> Since TestModule is not imported in AppModule in any way, adding or removing modules in TestModule should have no impact on main.js file size

That's totally wrong assumption, code splitting is based on module, not value, so that `@angular/core` can only exist in one bundle, in this case `main.js`. Any lazy bundle could increase the non-tree-shakable parts of `@angular/core`, resulting in bigger `@angular/core` remaining, thus bigger `main.js` bundle size.

As long as your lazy module make any additional usage of `@angular/core`, `@angular/common` of `rxjs`, size of `main.js` would certainly increased, and the stats seems expected so far.

**Comment by goodmite** [comment]
@trotyl  
Thanks for your reply. That's a big revelation for me. Can you refer to some doc?

However, don't you see that a bit problematic? Imagine your app having role-based access. Admin is having full access to all app while guest having very limited access. With the current build approach, the guest will also pay for the feature modules only meant for Admin (if those feature module use `@angular/common` etc).
